### ISSUE TITLE
Mixed system using real spaces

### ISSUE THREAD

**Issue by kalogirou** [firedrakeproject/firedrake#897]
When I run the code found here (https://bitbucket.org/annakalog/buoy2d/src/673fef3466b06c98ac11f5a120ff42117679d41e/Mixed%20system/?at=master), I get the following error. I believe it is related to real spaces. The firedrake version I am currently using is v0.13.0-477-g71cf933.

Traceback (most recent call last):
  File "buoy-swe.py", line 89, in <module>
    F_solver = solver_F(phi0_5, eta1, mu0_5, I, w, phi0, eta0, Z0, W0, step_b, etaR, phi_t, eta_t, mu_t, I_t, v1, v2, v3, v4, dt, Hb, H0, L, dR_dt, g, rho, Mass, solvers_print);
  File "/Users/matak/Documents/Simulations/Firedrake/Ship/Modules/Mixed system/solvers.py", line 41, in solver_F
    F_solver = LinearVariationalSolver(F_problem, solver_parameters=solvers_print)
  File "/Users/matak/firedrake/src/firedrake/firedrake/variational_solver.py", line 264, in __init__
    super(LinearVariationalSolver, self).__init__(*args, **kwargs)
  File "/Users/matak/firedrake/src/firedrake/firedrake/variational_solver.py", line 127, in __init__
    ctx = solving_utils._SNESContext(problem)
  File "/Users/matak/firedrake/src/firedrake/firedrake/solving_utils.py", line 107, in __init__
    for problem in problems)
  File "/Users/matak/firedrake/src/firedrake/firedrake/solving_utils.py", line 107, in <genexpr>
    for problem in problems)
  File "/Users/matak/firedrake/src/firedrake/firedrake/assemble.py", line 67, in assemble
    inverse=inverse, nest=nest)
  File "<decorator-gen-294>", line 2, in _assemble
  File "/Users/matak/firedrake/src/firedrake/firedrake/utils.py", line 62, in wrapper
    return f(*args, **kwargs)
  File "/Users/matak/firedrake/src/firedrake/firedrake/assemble.py", line 179, in _assemble
    nest=nest)
  File "/Users/matak/firedrake/src/PyOP2/pyop2/backends.py", line 118, in __call__
    return t(*args, **kwargs)
  File "/Users/matak/firedrake/src/PyOP2/pyop2/caching.py", line 122, in __new__
    args, kwargs = cls._process_args(*args, **kwargs)
  File "<decorator-gen-258>", line 2, in _process_args
  File "/Users/matak/firedrake/src/PyOP2/pyop2/utils.py", line 130, in wrapper
    return f(*args, **kwargs)
  File "/Users/matak/firedrake/src/PyOP2/pyop2/base.py", line 3591, in _process_args
    if not (pair[0].toset == dsets[0].set and
  File "/Users/matak/firedrake/src/PyOP2/pyop2/utils.py", line 64, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/Users/matak/firedrake/src/PyOP2/pyop2/base.py", line 3372, in toset
    m.toset for m in self._maps))
  File "/Users/matak/firedrake/src/PyOP2/pyop2/caching.py", line 156, in __new__
    obj = make_obj()
  File "/Users/matak/firedrake/src/PyOP2/pyop2/caching.py", line 137, in make_obj
    obj.__init__(*args, **kwargs)
  File "/Users/matak/firedrake/src/PyOP2/pyop2/base.py", line 977, in __init__
    "All components of a MixedSet must have the same number of layers."
AssertionError: All components of a MixedSet must have the same number of layers.

**Comment by colinjcotter** [comment]
Hi Anna,
  Please can you make a minimal test case, i.e. a simple script which exhibits the same problem, perhaps just assembling a mixed matrix with a real space in.

**Comment by kalogirou** [comment]
Hi Colin, here is a minimal example. Thank you!

```
from firedrake import *

# Parameter values
L = 1.0
H0 = 0.1
g = 9.81
r_M = 997.0/5
Nx = 160
a = atan(2.0/r_M/(L-0.75*L)**2)
dt = L/(Nx*2*pi*sqrt(g*H0))

# Define mesh and functions
m = IntervalMesh(Nx, L)
mesh = ExtrudedMesh(m, 10)

V = FunctionSpace(mesh, "CG", 1)
V_DG0 = FunctionSpace(mesh, "DG", 0)
V_R = FunctionSpace(mesh, "R", 0)
W = V*V*V*V_R

eta0 = Function(V)
phi0 = Function(V)
eta1 = Function(V)
phi0_5 = Function(V)
mu0_5 = Function(V)
I = Function(V_R)   # Function from the real space holding the integral of mu0_5
Z0 = Constant(0.0)
W0 = Constant(0.0)

w = Function(W)
phi, eta, mu, I_t = TrialFunctions(W)
v1, v2, v3, v4 = TestFunctions(W)
phi0_5, eta1, mu0_5, I = split(w)

# Wavemaker, buoy's hull shape and Heaviside step function
dR_dt = Function(V)
dR_dt.interpolate(Expression("A*sin(w*t)", A=0.2*(6*pi/Nx)*sqrt(g*H0), w=(6*pi/L)*sqrt(g*H0), t=0.5*dt))
Hb = Function(V)
Hb.interpolate(Expression("H0 - 0.5*(1.0+copysign(1.0,x[0]-0.75*L))*(sqrt(2*tan(a)/r_M)+tan(a)*(x[0]-L))", H0=H0, a=a, r_M=r_M, L=L))
step_b = Function(V_DG0)
step_b.interpolate(Expression("0.5*(1.0+copysign(1.0,x[0]-0.75*L))", L=L))

# Weak formulation
aphi = v1*(phi - step_b*mu)*dx
Lphi = v1*(phi0 - 0.5*dt*g*eta0)*dx
    
aeta = (v2*eta - dt*Hb*inner(grad(v2),grad(phi)))*dx
Leta = v2*eta0*dx + dt*H0*dR_dt*v2*ds_v(1)
    
aI = v4*(I_t/L-step_b*mu)*dx #I = assemble(step_b*mu*dx) = const. # I_t is the trial function
    
alambda = v3*step_b*(eta/dt + r_M*I_t)*dx
Llambda = v3*step_b*(Z0/dt + W0)*dx
    
a_solve = aphi + aeta + alambda + aI
L_solve = Lphi + Leta + Llambda
    
F_problem = LinearVariationalProblem(a_solve, L_solve, w) #, nest=False
F_solver = LinearVariationalSolver(F_problem, solver_parameters={'ksp_gmres_restart': 10000})

F_solver.solve()
```

**Comment by dham** [comment]
So, on current Firedrake I get a different error. There is a compiler error in the solver. The following is a significantly more minimal example. Interestingly, swapping the two lines indicated makes the error go away. This is because the error is the result of something going wrong in the way parloop arguments are mapped to Coefficients.

````python
from firedrake import *

parameters["pyop2_options"]["lazy_evaluation"] = False
parameters["assembly_cache"]["enabled"] = False

# Parameter values
Nx = 2
L = 1.

# Define mesh and functions
mesh = UnitSquareMesh(4, 4)

V = FunctionSpace(mesh, "CG", 1)
V_R = FunctionSpace(mesh, "R", 0)
W = V*V_R

eta,  I_t = TrialFunctions(W)
v2, v4 = TestFunctions(W)


Hb = Function(V) # Exchange these two lines to remove the error.
w = Function(W)


# Weak formulation

aeta = (Hb*dot(grad(v2),grad(eta)))*dx

aI = v4*(I_t)*dx

rhs = action(aeta + aI, w)

r=assemble(rhs)
Structure-preserving split finite element schemes for
geophysical applications
````

**Comment by dham** [comment]
OK I claim this is fixed in 9e9dc55

**Comment by kalogirou** [comment]
I can confirm that my original code now runs without errors.
